### PR TITLE
[11.0][IMP] hr_timesheet_sheet: fix amount calculation issues + code simplification

### LINKED ISSUE #198: [11.0][BUG] hr_timesheet_sheet: unable to remove all lines of a timesheet sheet
Reproduced in Runbot:

 - Create some lines in the details tab. Save.
 - Click on Edit.
 - In the details tab, remove the lines one by one (click on the trash bin icon).

When attempting to remove the last line, all the previously deleted lines reappear again.

### LINKED ISSUE #209: [11.0][BUG] hr_timesheet_sheet: Details tab not updated when changing Employee
Create a timesheet sheet:

```
Employee: Pieter Parker
Timesheet Period: from 03/18/2019 to 03/24/2019 (Week 12)

```
Add a line:

```
03/18/2019	Data Import/Export Plugin		Test1	01:00
```

Save. The data in the Summary tab is now aligned with Details tab.

Edit and change Employee: for example choose `Demo User`. Save.

Result:
Summary tab is empty (ok). Details tab is not updated and still displays the line (not ok).

Expected:
Details tab should be updated.

### PR DESCRIPTION
**Pull request by astirpe**
This PR fixes the amount calculation issues (see https://github.com/OCA/hr-timesheet/issues/206) regarding the decrease of the amount of hours from the summary tab.
The idea is that decreasing the amount from the summary tab should not directly decrease the amounts of time for existing projects and/or tasks, to avoid potential loose of relevant information. In case it's needed, it's up to the end user to manually adjust the amounts or merge lines in the details tab.
This is basically achieved by removing the `def _diff_amount_timesheets()` method.

The proposed solution also causes a code reduction (simplified code) including the following:

 - Removed `count_timesheets` field
 - Simplified code for `def onchange_unit_amount()`
 - Makes use of field `sheet.timesheet_ids` in place of `self.env['account.analytic.line'].search()`

This PR implicitly allows negative amounts: [two lines](https://github.com/OCA/hr-timesheet/blob/11.0/hr_timesheet_sheet/models/hr_timesheet_sheet.py#L578-L579) are removed from `def onchange_unit_amount()`, the same way as it is done in https://github.com/OCA/hr-timesheet/pull/201.

This PR also fixes method `def _default_employee()` in a multicompany environment, filtering the default employee by company.

Fixes #198 

Fixes #209

### CODE REVIEW

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 1]
```diff
@@ -257,10 +257,11 @@ def _compute_line_ids(self):
             dates = sheet._get_dates()
             if not dates:
                 continue
-            timesheets = sheet._get_timesheet_lines()
             lines = self.env['hr_timesheet.sheet.line']
             for date in dates:
-                for project in timesheets.mapped('project_id'):
+                timesheets = sheet._get_timesheet_lines()  # do not move this
```
Comment (MiquelRForgeFlow) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: why?

**Review by MiquelRForgeFlow — CHANGES_REQUESTED** [review]
> This PR implicitly includes a fix for the negative amounts. The solution is a bit alternative to #201.

That's not a fix for the negative amounts. It's just the removal of the non negative amounts condition as in #201. And then, why do you propose an alternative? What's wrong with the #201? Now, will reviewers need to choose between two PRs? I don't understand. I think you could propose here only the other changes (the ones solving your issue #206) instead of doing also commits that enter in conflict with #201. If you want changes, then request changes in my PR, but don't open alternative PRs regarding a same topic (negative amounts). Also, I could be happy to include the solution of your issue in #201, it's easy.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 1 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Otherwise you will get errors on tests and while using the module.

**Comment by astirpe** [comment]
@mreficent While simplifying that method, I just deleted the same 2 lines you already deleted in your PR. That was necessary because of the design of this PR, the reason is already explained above: this PR fixes the decreasing of the amounts, and decreasing by setting negative amounts is also part of the cases fixed by this PR. Nobody is complaining about your PR https://github.com/OCA/hr-timesheet/pull/201 which remains valid.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 1 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Well, non blocking, you should know that this will be just a temporal fix. It's not a correct fix, because you are doing a search for each date and it's inefficient, but just serves to bypass the errors produced by the other changes. This fix it's related to merged timesheets and having deleted records in `self.timesheet_ids`. In https://github.com/OCA/hr-timesheet/pull/188, when applying https://github.com/OCA/hr-timesheet/pull/188/commits/8ab95b7af177608ed055b70fc3a87088d6d03d01, then is consequently necessary apply https://github.com/OCA/hr-timesheet/pull/188/commits/3e7c2c78b8bd8551544b94134e19351c69f50865 where the correct fix applies.

**Comment by MiquelRForgeFlow** [comment]
@astirpe ok, I will check

**Comment by astirpe** [comment]
Commit https://github.com/OCA/hr-timesheet/pull/207/commits/b5d1a7a7b08e73246a144b37a269fb5c1042bd40 should fix the issue mentioned in https://github.com/OCA/hr-timesheet/pull/207#discussion_r265217274

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 2]
```diff
@@ -271,16 +271,30 @@ def _compute_line_ids(self):
...
-                                date=date,
-                                project=project,
-                                task=task,
-                                timesheets=timesheet.filtered(
-                                    lambda t: date == t.date
-                                    and t.task_id.id == task.id),
-                            ))
-            sheet.line_ids = lines
+                        data_key = (date, project, task)
+                        timesheet = timesheet.filtered(
```
Comment (MiquelRForgeFlow) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: `timesheet = timesheet_pr.filtered(`

where `timesheet_pr` should be the one created in https://github.com/OCA/hr-timesheet/pull/207/commits/b5d1a7a7b08e73246a144b37a269fb5c1042bd40#diff-a48ce9b933da452cb1910c0e256932c7R267.

File: hr_timesheet_sheet/__manifest__.py [hunk 3]
```diff
@@ -9,7 +9,7 @@
     'sequence': 80,
     'summary': 'Timesheet Sheets, Activities',
     'license': 'AGPL-3',
-    'author': 'Eficent, Odoo Community Association (OCA)',
+    'author': 'Eficent, Onestein, Odoo Community Association (OCA)',
```
Comment (nikul-serpentcs) [inline hr_timesheet_sheet/__manifest__.py:12]: @astirpe Don't change the author

**Review by nikul-serpentcs — CHANGES_REQUESTED** [review]
Minor change

File: hr_timesheet_sheet/__manifest__.py (hunk 3 above)
Comment (astirpe) [reply hr_timesheet_sheet/__manifest__.py:12]: Why not?
I spent a lot of effort in this module, see all the related PRs, issues, reviews, functional tests, discussions, etc... and all of this was possible thanks to Onestein.

Comment (nikul-serpentcs) [reply hr_timesheet_sheet/__manifest__.py:12]: @astirpe As per my knowledge, you can not change `author`, You can add your name in the `contribution section `.

Sorry when I am Wrong!

@pedrobaeza @jbeficent Can you please brief here

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 4]
```diff
@@ -575,83 +566,35 @@ def onchange_unit_amount(self):
...
-                        else:
-                            self._diff_amount_timesheets(diff_amount, other_ts)
-                else:
-                    raise ValidationError(
-                        _('Error code: Cannot have 0 timesheets.'))
+
+        sheet = self.sheet_id
+        if not sheet:
+            model = self.env.context.get('params').get('model', '')
+            obj_id = self.env.context.get('params').get('id')
```
Comment (MiquelRForgeFlow) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: why? could you explain this?

File: hr_timesheet_sheet/__manifest__.py (hunk 3 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/__manifest__.py:12]: I, as the original author, I think @astirpe has contributed a lot in this module, sufficiently to be recognized him as another author. So, no problem from me with sharing authorship, nor any complain from @Eficent :+1:

**Review by MiquelRForgeFlow — APPROVED** [review]
Review completed. Good changes. :+1: 

But please, answer my question above

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 2 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: this it's resolved, it can be minimized

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 1 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: this can be minimized

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 4 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Of course. Is for when you create a new timesheet sheet not yet saved (not stored on the DB) and you start to change the amounts in the matrix.

The following is an example of what happens without that block of code.
I invite you to reproduce the steps an debug the code: it's the best way to understand this part IMO.



Be sure that following lines already exist without any created sheet:
```
03/13/2019    Data Import/Export Plugin        test1    01:00   
03/13/2019    Data Import/Export Plugin        test2    02:00   
```

(You can also create the lines from a timesheet sheet and then delete the sheet.)


Create an empty timesheet sheet but do not save yet.

In Summary tab, change value from 3.0 to 4.0. Save.

This is what happens:

```
03/13/2019	Data Import/Export Plugin	test1	01:00
03/13/2019	Data Import/Export Plugin	test2	02:00
03/13/2019	Data Import/Export Plugin	/	04:00
```

This is what is expected:

```
03/13/2019	Data Import/Export Plugin	test1	01:00
03/13/2019	Data Import/Export Plugin	test2	02:00
03/13/2019	Data Import/Export Plugin	/	01:00
```

Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Oh, right, I understand :+1:  Yeah, that was a headache.

good to learn this magic 🧙‍♂️ my spell book is increasing

**Comment by MiquelRForgeFlow** [comment]
@alexey-pelykh could you review this?

**Comment by astirpe** [comment]
Commit [[FIX] Allow to delete all lines + better tab sync ](https://github.com/OCA/hr-timesheet/pull/207/commits/0d12ff72e02eaf353f90939258c23cf1e7b24cf8) contains a fix for issue https://github.com/OCA/hr-timesheet/issues/198. The fix is better than the one proposed in https://github.com/OCA/hr-timesheet/pull/203.

The proposed solution consists in splitting and fixing the onchange method of the date fields and the `timesheet_ids`.

As a consequence of fixing the onchange methods, another wrong behavior is fixed: when deleting lines in the Details tab, the Summary tab is automatically updated (it was not updated before this commit).

Supersedes https://github.com/OCA/hr-timesheet/pull/203.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 5]
```diff
@@ -323,12 +317,6 @@ def onchange_add_project_id(self):
                          self.timesheet_ids.mapped('task_id').ids)],
                 },
             }
-        else:
-            return {
```
Comment (MiquelRForgeFlow) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: why?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 6]
```diff
@@ -109,14 +110,16 @@ def _default_employee(self):
         string='Timesheets',
         readonly=True,
         states={
+            'new': [('readonly', False)],
             'draft': [('readonly', False)],
         }
     )
     state = fields.Selection([
+        ('new', 'New'),
```
Comment (MiquelRForgeFlow) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py:118]: why?

**Comment by astirpe** [comment]
Commit https://github.com/OCA/hr-timesheet/pull/207/commits/455d28014be412dbcf8602e812484a085796ee4a fixes the onchange of field employee_id (fixes https://github.com/OCA/hr-timesheet/issues/209).

The proposed fix restores the state 'New' for the timesheet sheet, like it was in standard Odoo 10.0 (see [line](https://github.com/odoo/odoo/blob/10.0/addons/hr_timesheet_sheet/models/hr_timesheet_sheet.py#L62)).
 
Restoring the state 'New' is necessary to impede the onchange of `date_start`, `date_end` and `employee_id` being triggered when the status is "Draft". In facts performing the onchange when the status is "Draft" will cause the timesheet lines being deleted from database (this must be avoided!).

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 5 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Because `add_line_task_id` is invisible when `add_line_project_id` is empty, so this part of code is not reachable.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 6 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:118]: See https://github.com/OCA/hr-timesheet/pull/207#issuecomment-474303424

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 5 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Invisible doesn't mean empty. It could be wrongly filled and still be invisible. Better assure the domain independently if it is invisible or not just in case.

Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: :thinking: why do you need to set a different domain for a field that is invisible?

Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: I don't understand your point, sorry.

Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Because maybe you have an extension where you don't have invisible that field.

Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Ok, I'm going to restore that block of code then :+1:

Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Well, yeah, don't worry. Resolve and minimize this observation.

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 6 above)
Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:118]: I understand :+1:

Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py:118]: Resolved

**Review by MiquelRForgeFlow — APPROVED** [review]

**Comment by astirpe** [comment]
Squashed and rebased

File: hr_timesheet_sheet/models/hr_employee.py [hunk 7]
```diff
@@ -17,7 +17,7 @@ class HrEmployee(models.Model):
     @api.multi
     def _compute_timesheet_count(self):
         for employee in self:
-            employee.timesheet_count = employee.env['hr_timesheet.sheet'].\
+            employee.timesheet_count = self.env['hr_timesheet.sheet'].\
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_employee.py]: This may be extracted as `Sheet = self.env['hr_timesheet.sheet']` outside the loop

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 8]
```diff
@@ -257,47 +258,53 @@ def _compute_line_ids(self):
...
-            timesheets = sheet._get_timesheet_lines()
-            lines = self.env['hr_timesheet.sheet.line']
+            data_matrix = {}
+            projects = sheet.timesheet_ids.mapped('project_id')
             for date in dates:
-                for project in timesheets.mapped('project_id'):
-                    timesheet = timesheets.filtered(
+                for project in projects:
+                    timesheet_pr = sheet.timesheet_ids.filtered(
                         lambda x: (x.project_id == project))
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: ( and ) are not needed

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 9]
```diff
@@ -257,47 +258,53 @@ def _compute_line_ids(self):
...
                 continue
-            timesheets = sheet._get_timesheet_lines()
-            lines = self.env['hr_timesheet.sheet.line']
+            data_matrix = {}
+            projects = sheet.timesheet_ids.mapped('project_id')
             for date in dates:
-                for project in timesheets.mapped('project_id'):
-                    timesheet = timesheets.filtered(
+                for project in projects:
+                    timesheet_pr = sheet.timesheet_ids.filtered(
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: `project_timesheets` maybe?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 10]
```diff
@@ -257,47 +258,53 @@ def _compute_line_ids(self):
...
-                                project=project,
-                                task=task,
-                                timesheets=timesheet.filtered(
-                                    lambda t: date == t.date
-                                    and t.task_id.id == task.id),
-                            ))
-            sheet.line_ids = lines
-
-    def _get_timesheet_lines(self):
+                        data_key = (date, project, task)
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Would it be beneficial to use named tuple?

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 11]
```diff
@@ -257,47 +258,53 @@ def _compute_line_ids(self):
...
+                    date=item[0],
+                    project=item[1],
+                    task=item[2],
+                    unit_amount=data_matrix[item]['unit_amount']
+                ))
+            self.clean_timesheets(data_matrix[item]['timesheets'])
+        return lines
+
+    @api.onchange('date_start', 'date_end', 'employee_id')
+    def onchange_dates(self):
```
Comment (alexey-pelykh) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: I've seen `_onchange_` prefix more, as it's "private" method

**Review by alexey-pelykh — CHANGES_REQUESTED** [review]
Did the code review, yet no functional test yet

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py (hunk 10 above)
Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: The only benefit I see is the elegance of code but I think it doesn't worth it.

**Review by alexey-pelykh — APPROVED** [review]
Code LGTM, yet I'm already scared how this going to be ported to v12

**Comment by alexey-pelykh** [comment]
@astirpe please squash the code review commit as fixup into other commit

**Comment by astirpe** [comment]
@alexey-pelykh I can do the porting to V12, I think I'm now confident enough with this module.

**Comment by alexey-pelykh** [comment]
@astirpe That would be much appreciated!

**Comment by astirpe** [comment]
Porting to V12: https://github.com/OCA/hr-timesheet/pull/211

**Comment by astirpe** [comment]
I need also https://github.com/OCA/hr-timesheet/pull/207/commits/412a80167a6f9d8f0d7f01458fef7159092054d7 to make customization easier

**Comment by astirpe** [comment]
And https://github.com/OCA/hr-timesheet/pull/207/commits/4dc60939393afd42a68bbe7b34ed1d3a67dce925 makes the list view more readable.

**Comment by astirpe** [comment]
Commit https://github.com/OCA/hr-timesheet/pull/207/commits/b2c5b305c3195eb83f555645e532e1b540bda81c simplifies the code and I think also with performance benefits

File: hr_timesheet_sheet/models/hr_timesheet_sheet.py [hunk 12]
```diff
@@ -287,7 +287,7 @@ def _create_data_matrix_lines(self, data_matrix):
                     unit_amount=data_matrix[item]['unit_amount']
                 ))
             self.clean_timesheets(data_matrix[item]['timesheets'])
-        return lines
+        return lines.sorted(key=lambda l: l.date)
```
Comment (MiquelRForgeFlow) [inline hr_timesheet_sheet/models/hr_timesheet_sheet.py]: and also project, perhaps? :thinking:

Comment (MiquelRForgeFlow) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: and also task hahahaha

Comment (alexey-pelykh) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: I'd extract the key to a function for extending

Comment (astirpe) [reply hr_timesheet_sheet/models/hr_timesheet_sheet.py]: Now `test_6` is angry because of the sorting :thinking:

**Comment by astirpe** [comment]
Commit https://github.com/OCA/hr-timesheet/pull/207/commits/66a2455f528ac6767fc7591bc8141771a91b95ff fixes the following issue.

1. Be sure that admin user belongs to company: YourCompany.
2. Create CompanyB.
3. Create UserB, belonging to CompanyB.
4. Create EmployeeB, belonging to CompanyB and linked to UserB.
5. Login as admin and create a new timesheet sheet. Do not save yet.
6. Change employee to EmployeeB.

Company is not updated: it stays YourCompany (not ok).
It is not possible to change the Company because it's a readonly field.
Save. Error is displayed: "The Company in the Timesheet Sheet and in the Employee must be the same".

**Review by MiquelRForgeFlow — APPROVED** [review]

**Comment by alexey-pelykh** [comment]
@astirpe do you plan other changes in this PR?

**Comment by astirpe** [comment]
@alexey-pelykh No. I would be happy to stop it here. :wink:

**Comment by alexey-pelykh** [comment]
@nikul-serpentcs please re-review

**Review by nikul-serpentcs — APPROVED** [review]
Only Code review LGTM :+1:

**Comment by alexey-pelykh** [comment]
@pedrobaeza please take a look at this PR

**Comment by pedrobaeza** [comment]
Merging blindly on my side due to previous approvals.
